HPA panics

I just run into panic in controller-manager:
```
/usr/local/go/src/runtime/asm_amd64.s:2086
panic: runtime error: integer divide by zero [recovered]
        panic: runtime error: integer divide by zero [recovered]
        panic: runtime error: integer divide by zero

goroutine 327 [running]:
panic(0x229cb80, 0xc420010040)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kubernetes/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:56 +0x126
panic(0x229cb80, 0xc420010040)
        /usr/local/go/src/runtime/panic.go:458 +0x243
k8s.io/kubernetes/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:56 +0x126
panic(0x229cb80, 0xc420010040)
        /usr/local/go/src/runtime/panic.go:458 +0x243
k8s.io/kubernetes/pkg/controller/podautoscaler/metrics.GetResourceUtilizationRatio(0xc4218cb1a0, 0xc421064ce0, 0xc40000003c, 0x26, 0xc4217c7408, 0x38cc801, 0xc421995740)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/metrics/utilization.go:37 +0x196
k8s.io/kubernetes/pkg/controller/podautoscaler.(*ReplicaCalculator).GetResourceReplicas(0xc420907f60, 0x3c00000002, 0x26fdc1a, 0x3, 0xc42090ae54, 0x7, 0x38cc8e0, 0xc421995740, 0xc4212da480, 0x26ffbca, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/replica_calculator.go:97 +0x958
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).computeReplicasForCPUUtilization(0xc4208fbc70, 0xc4203d2f08, 0xc4212da480, 0x37, 0x39460e0, 0xc4212da400, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:167 +0x3e1
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).reconcileAutoscaler(0xc4208fbc70, 0xc4203d2f08, 0x27855ae, 0x37)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:315 +0xc42
k8s.io/kubernetes/pkg/controller/podautoscaler.newInformer.func3(0x26b07a0, 0xc4203d2f08)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:95 +0xc3
k8s.io/kubernetes/pkg/client/cache.ResourceEventHandlerFuncs.OnAdd(0xc4208d2200, 0xc4208d2210, 0x0, 0x26b07a0, 0xc4203d2f08)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:173 +0x49
k8s.io/kubernetes/pkg/client/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc4208fe0c0, 0x26b07a0, 0xc4203d2f08)
        <autogenerated>:50 +0x78
k8s.io/kubernetes/pkg/client/cache.NewInformer.func1(0x22f5240, 0xc4219952e0, 0xc4219952e0, 0x22f5240)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:251 +0x270
k8s.io/kubernetes/pkg/client/cache.(*DeltaFIFO).Pop(0xc420864630, 0xc42087bd10, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/delta_fifo.go:420 +0x22a
k8s.io/kubernetes/pkg/client/cache.(*Controller).processLoop(0xc42018ae00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:131 +0x3c
k8s.io/kubernetes/pkg/client/cache.(*Controller).(k8s.io/kubernetes/pkg/client/cache.processLoop)-fm()
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:102 +0x2a
k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1(0xc421065f60)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:96 +0x5e
k8s.io/kubernetes/pkg/util/wait.JitterUntil(0xc421065f60, 0x3b9aca00, 0x0, 0x2164401, 0xc420126a20)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:97 +0xad
k8s.io/kubernetes/pkg/util/wait.Until(0xc421065f60, 0x3b9aca00, 0xc420126a20)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:52 +0x4d
k8s.io/kubernetes/pkg/client/cache.(*Controller).Run(0xc42018ae00, 0xc420126a20)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/controller.go:102 +0x1af
created by k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).Run
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:133 +0xc0
```

This clearly shouldn't happen.

cc @fgrzadkowski @jszczepkowski @kubernetes/sig-autoscaling-misc

--- a/pkg/controller/podautoscaler/metrics/utilization.go
+++ b/pkg/controller/podautoscaler/metrics/utilization.go
@@ -16,6 +16,10 @@ limitations under the License.
 
 package metrics
 
+import (
+	"fmt"
+)
+
 // GetResourceUtilizationRatio takes in a set of metrics, a set of matching requests,
 // and a target utilization percentage, and calcuates the the ratio of
 // desired to actual utilization (returning that and the actual utilization)
@@ -32,6 +36,12 @@ func GetResourceUtilizationRatio(metrics PodResourceInfo, requests map[string]in
 
 		metricsTotal += metricValue
 		requestsTotal += request
+	}
+
+	// if the set of requests is completely disjoint from the set of metrics,
+	// then we could have an issue where the requests total is zero
+	if requestsTotal == 0 {
+		return 0, 0, fmt.Errorf("no metrics returned matched known pods")
 	}
 
 	currentUtilization := int32((metricsTotal * 100) / requestsTotal)

--- a/pkg/controller/podautoscaler/replica_calculator_test.go
+++ b/pkg/controller/podautoscaler/replica_calculator_test.go
@@ -47,6 +47,8 @@ type resourceInfo struct {
 	name     v1.ResourceName
 	requests []resource.Quantity
 	levels   []int64
+	// only applies to pod names returned from "heapster"
+	podNames []string
 
 	targetUtilization   int32
 	expectedUtilization int32
@@ -134,9 +136,13 @@ func (tc *replicaCalcTestCase) prepareTestClient(t *testing.T) *fake.Clientset {
 		if tc.resource != nil {
 			metrics := metricsapi.PodMetricsList{}
 			for i, resValue := range tc.resource.levels {
+				podName := fmt.Sprintf("%s-%d", podNamePrefix, i)
+				if len(tc.resource.podNames) > i {
+					podName = tc.resource.podNames[i]
+				}
 				podMetric := metricsapi.PodMetrics{
 					ObjectMeta: v1.ObjectMeta{
-						Name:      fmt.Sprintf("%s-%d", podNamePrefix, i),
+						Name:      podName,
 						Namespace: testNamespace,
 					},
 					Timestamp: unversioned.Time{Time: tc.timestamp},
@@ -248,6 +254,22 @@ func (tc *replicaCalcTestCase) runTest(t *testing.T) {
 		assert.InDelta(t, tc.metric.expectedUtilization, 0.1, outUtilization, "utilization should be as expected")
 		assert.True(t, tc.timestamp.Equal(outTimestamp), "timestamp should be as expected")
 	}
+}
+
+func TestReplicaCalcDisjointResourcesMetrics(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas: 1,
+		expectedError:   fmt.Errorf("no metrics returned matched known pods"),
+		resource: &resourceInfo{
+			name:     v1.ResourceCPU,
+			requests: []resource.Quantity{resource.MustParse("1.0")},
+			levels:   []int64{100},
+			podNames: []string{"an-older-pod-name"},
+
+			targetUtilization: 100,
+		},
+	}
+	tc.runTest(t)
 }
 
 func TestReplicaCalcScaleUp(t *testing.T) {

